### DISCUSSION TITLE
PlayerInteractHandEvent and PlayerInteractBlockEvent

### DISCUSSION THREAD

**Discussion by NonSwag** [PaperMC/Paper#11207] [Ideas]
### Is your feature request related to a problem?

If you want to add actions to items or blocks when interacting with them (right or left click) you have to do a shtload of code

# Example - Item action

Adding an interaction to a stick when held in any hand **_but not both at the same time_**
which only works when you don't look at an intractable block, would require following:

0. listen to right-clicks on air and blocks
1. check which block is clicked
   1. maintain a list of actual intractable blocks
   2. perform special checks based on the block whether it will perform any action based on conditions (will the block open an interface, can it be used without a correct item, will the item be consumed, is it on cooldown...)
   3. ignore the event for blocks with interactions
2. check which hand the item is in
   1. check main hand for the item: ignore if hand in event was offhand
   2. check offhand for the item: ignore if hand in event was main hand
3. perform your action
 
# Example - Block action

Adding an interaction to a block with an item would require the following:

0. listen to right-click on blocks
1. check if the item has an interaction
   1. maintain a list of actual intractable items
   2. perform special checks based on the item whether it will perform any action based on conditions (will the block place, will the item be consumed, will the item be launched, is the item on cooldown...)
   3. ignore the event for items that perform an action
2. all the points of the item action example apply again
3. perform your action

### Describe the solution you'd like.

I would propose two new events for differentiating block and hand interactions
The events could be called: PlayerInteractHandEvent and PlayerInteractBlockEvent
The interact hand event would provide the hand and item used and the default event result
The interact block event would provide the block and block face as well as the hand and the default event result

### Describe alternatives you've considered.

Currently, you have to check the clicked block and perform your own logic to get the correct item used inside an interaction only to determine which hand was used and if the interaction will perform something on the item or in the world, which is quite tidies

**Comment by NonSwag** [comment]
Would also kinda solve #11181

**Comment by electronicboy** [comment]
Not sure this is really tenable, if you right click a block, vanilla always tries to interact with that block, and then falls back to using the item, we don't know what the interaction result will be until we run said interaction

**Reply by NonSwag** [reply]
Would it be feasible to split the checks and actions into separate methods
For example:
`getInteractionResult(ServerPlayer,Item)`
`performAction(ServerPlayer,Item)` 

In the interaction logic:
get the interaction result for the player using that item
If success call perform action

I think the maintenance effort would be huge but on the other hand unclutter the code immensely and make the lives of many developers way easier
Custom item/block interactions would be easier to implement 
Also when working with internals add more fine control to things
As well as mitigating potential side effects when getting a value that does not communicate it performs any action

No idea if this text even makes sense
_Written on my phone during lunchtime_
